storage/replicate_queue.go

```diff
@@ -99,28 +113,45 @@ func (rq replicateQueue) process(now proto.Timestamp, repl *Replica) error {
 	}
 
 	desc := repl.Desc()
-	if needs, _ := rq.needsReplication(zone, repl, desc); !needs {
+	delta := rq.replicaDelta(zone, repl, desc)
+	if delta == 0 {
 		// Something changed between shouldQueue and process.
 		return nil
 	}
 
-	// TODO(bdarnell): handle non-homogeneous ReplicaAttrs.
-	// Allow constraints to be relaxed if necessary.
-	newReplica, err := rq.allocator.allocateTarget(zone.ReplicaAttrs[0], desc.Replicas, true, nil)
-	if err != nil {
-		return err
-	}
+	// TODO(bdarnell): handle non-homogenous ReplicaAttrs.
+	if delta > 0 {
+		// Allow constraints to be relaxed if necessary.
+		newReplica, err := rq.allocator.allocateTarget(zone.ReplicaAttrs[0], desc.Replicas, true, nil)
+		if err != nil {
+			return err
+		}
 
-	replica := proto.Replica{
-		NodeID:  newReplica.Node.NodeID,
-		StoreID: newReplica.StoreID,
-	}
-	if err = repl.ChangeReplicas(proto.ADD_REPLICA, replica, desc); err != nil {
-		return err
+		replica := proto.Replica{
+			NodeID:  newReplica.Node.NodeID,
+			StoreID: newReplica.StoreID,
+		}
+		if err = repl.ChangeReplicas(proto.ADD_REPLICA, replica, desc); err != nil {
+			return err
+		}
+	} else {
+		removeReplica, err := rq.allocator.removeTarget(desc.Replicas)
+		if err != nil {
+			return err
+		}
+		if log.V(1) {
+			log.Infof("Remove replica %s", removeReplica)
+		}
+
+		if err = repl.ChangeReplicas(proto.REMOVE_REPLICA, removeReplica, desc); err != nil {
+			if log.V(1) {
+				log.Infof("Error removing replica %s", err)
+			}
+			return err
+		}
 	}
 
 	// Enqueue this replica again to see if there are more changes to be made.
-	//go rq.MaybeAdd(repl, rq.clock.Now())
```

Comment:
I'm going to leave the `go` out, I don't know why it would be there in the first place (unless, as Ben said, there was a re-entrant lock of some kind).